### PR TITLE
Network: use device name in the host to determine nic's type

### LINKED ISSUE #981: Potential bug when determining nic's type in AddInterface
# Description of problem

We reuse the logic of createEndpointsFromScan() in the AddInterface() to determine a nic's type. While originally it scans a network ns, finds a link and generate NetworkInfo from link. It will use link's name to determine whether it is a physical/vhost_user nic.
Now physical and vhost_user endpoints hotplug are not supported yet. But in AddInterface(), we generate NetworkInfo from cli args. The name field is actually what we want to set into the VM. So it is never existed in the host. This may be a potential bug. I think we should use device field instead?

### PR DESCRIPTION
**Pull request by caoruidong**
When adding interface, we should use device field in API to determine a
nic's type, not the name we set into the VM.

Fixes #981

Signed-off-by: Ruidong Cao <caoruidong@huawei.com>

### CODE REVIEW

**Comment by bergwolf** [comment]
@caoruidong Does it fix the issue by just setting `Iface.Name` to `inf.Device` in `generateNetInfo` instead? Unlike `types.Interface`, `NetworkInfo` is the runtime internal data structure where the `Iface.Name` field refers to the host network device. /cc @amshinde

File: virtcontainers/sandbox.go [hunk 1]
```diff
@@ -1111,9 +1112,42 @@ func (s *Sandbox) generateNetInfo(inf *types.Interface) (NetworkInfo, error) {
...
-	netInfo, err := s.generateNetInfo(inf)
+	var netInfo NetworkInfo
+
+	netlinkHandle, err := getNetlinkHandle(s.config.NetworkConfig.NetNSPath)
+	if err != nil {
+		return nil, err
+	}
+	defer netlinkHandle.Delete()
+
+	link, err := netlinkHandle.LinkByName(inf.Device)
```
Comment (jodh-intel) [inline virtcontainers/sandbox.go]: I'm wondering if we should add a check for `inf != nil`?

**Comment by caoruidong** [comment]
@bergwolf I observe that Device field is actually not used ether in runtime or agent. So I'm not very clear about its semantic meaning. I think your advice is a good one. But Iface.Name may be used somewhere else. I will go through the whole process to check.

**Comment by bergwolf** [comment]
LGTM. Thanks @caoruidong !

**Comment by jodh-intel** [comment]
@caoruidong - this change looks harmless, but I don't understand it as "it looks wrong" :(

Also, it's these sorts of tiny changes which can turn into nasty bugs. This PR might be fixing a bug, but I'd feel more comfortable if there was atleast one unit test with this change :)

@amshinde, @mcastelino, @sboeuf - could you clarify some of this?

**Comment by caoruidong** [comment]
@jodh-intel Yes, I agree. I suspect it is a potential bug when hot attaching a physical nic. Though this feature has not been merged. I'll see if I can add one test

**Comment by raravena80** [comment]
@caoruidong any updates on this PR?

**Comment by caoruidong** [comment]
https://mvdan.cc/xurls/cmd/xurls is broken again

**Comment by caoruidong** [comment]
This PR is not necessary. I just want to record a potential issue.
Now the device field is never used except logging.  I think we need to clarify the purpose of every field when hotplug different types of nic like physical and veth.

**Pull request closed without merging**